compiler/rustc_lint/src/internal.rs

```diff
@@ -122,7 +125,121 @@ impl LateLintPass<'_> for QueryStability {
                 );
             }
         }
+        check_into_iter_stability(cx, expr);
+    }
+}
+
+fn check_into_iter_stability<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+    let Some(into_iterator_def_id) = cx.tcx.get_diagnostic_item(sym::IntoIterator) else {
+        return;
+    };
+    if expr.span.from_expansion() {
+        return;
+    };
+    // Is `expr` a function or method call?
+    let Some((callee_def_id, generic_args, recv, args)) =
+        get_callee_generic_args_and_args(cx, expr)
+    else {
+        return;
+    };
+    let fn_sig = cx.tcx.fn_sig(callee_def_id).instantiate_identity().skip_binder();
+    for (arg_index, &input) in fn_sig.inputs().iter().enumerate() {
+        let &ty::Param(ParamTy { index: param_index, .. }) = input.kind() else {
```

Comment:
Right, this has a bunch of false negatives. E.g., it won't consider `Option<T>`, `(impl IntoIterator,)` etc. but I guess it's still better than the status quo.

You could create a custom `TypeVisitor` to find all containing type params in a `Ty` but that's probably overkill.